### PR TITLE
Early stop training if model weights have nan or inf tensors

### PR DESCRIPTION
**Pull request by arnavgarg1**
# What does this PR do?

This PR introduces a new early stopping behavior when NaN or Inf values are detected in the model weights. This is very common during fine-tuning of LLMs since fine-tuning is a relatively unstable process. The new mechanism:

1. Waits for training to finish for the current epoch (or until steps_per_checkpoint or train_steps are reached).
2. Checks if the models weights or buffers have NaN or Inf values. 
  - If they are not NaN or Inf, training continues to evaluation where the model is evaluated on the validation and test sets, and if the validation metric improves, the new weights are saved and the progress tracker is updated.
  - If they are not NaN or Inf, training stops. The current model weights are not saved as a checkpoint and the training progress tracker is not updated. Instead, training terminates early by breaking out of the training loop. The best checkpoint's weights are loaded into the model for inference. This is most likely to be the weights of the previous, valid checkpoint, but will be the last valid checkpoint with the best validation metric score. 
  _Note: In the instance that the first checkpoint itself is NaN, a RuntimeError is raised indicating that training failed because no valid checkpoint was saved._
3. You can now use this for inference.

I've verified that this works without running into errors in the following scenarios ✅ :
1. LLM fine-tuning using local backend on a single GPU
2. LLM fine-tuning using local backend on multiple GPUs (naive sharding/parallelism)
3. LLM fine-tuning using ray backend with deepspeed on multiple workers
4. ECD with DDP on multiple workers

---
## Example output

I fine-tuned a model using a config where the first epoch trains successfully without NaN loss or weights, but the second epoch runs into NaN loss because of NaN weights.

This is the output from early stopping because of NaN weight detection. Note:
1. It runs this check before running evaluation after the 2nd epoch of training.
6. It tells you that the first LoRA layer had NaN weights and that it is stopping training.
7. Model training stops and does not continue.

```
Training:  38%|██████████████████████████████████████████▎                                                                     | 1077/2850 [11:26<17:58,  1.64it/s, loss=nan]
/home/ray/anaconda3/lib/python3.8/site-packages/torchmetrics/utilities/prints.py:43: UserWarning: Encountered `nan` values in tensor. Will be removed.
  warnings.warn(*args, **kwargs)  # noqa: B028
Training:  67%|██████████████████████████████████████████████████████████████████████████▋                                     | 1900/2850 [19:36<09:30,  1.66it/s, loss=nan]
Found NaN or inf values in parameter 'model.base_model.model.model.layers.0.self_attn.q_proj.lora_A.default.weight' of module 'LLM'
NaN or inf tensors found in the model. Stopping training.
Training:  67%|██████████████████████████████████████████████████████████████████████████▋                                     | 1900/2850 [19:36<09:48,  1.62it/s, loss=nan]

╒══════════╕
│ FINISHED │
╘══════════╛
```

We can then load the model back in and check the same layer to make sure that the saved weights were from the previous checkpoint and aren't NaN:

```
>>> model.model.model.base_model.model.model.layers[0].self_attn.q_proj.lora_A.default.weight
Parameter containing:
tensor([[ 0.0086,  0.0118,  0.0147,  ...,  0.0103, -0.0025,  0.0163],
        [-0.0061,  0.0095, -0.0034,  ..., -0.0106, -0.0164,  0.0098],
        [-0.0012,  0.0127,  0.0052,  ...,  0.0132,  0.0090, -0.0094],
        ...,
        [-0.0086,  0.0032,  0.0125,  ...,  0.0147,  0.0177,  0.0139],
        [ 0.0131, -0.0035,  0.0117,  ...,  0.0169,  0.0005, -0.0142],
        [ 0.0012, -0.0119, -0.0082,  ..., -0.0078, -0.0034, -0.0138]],
       device='cuda:0')
```

Finally, we can try and run inference to make sure that it does not produce an error:

```
predictions = model.predict(test_examples, generation_config={"temperature": 0.1, "max_new_tokens": 128})[0]
>>> for input_with_prediction in zip(test_examples["comment_text"], predictions["is_toxic_response"]):
...     print(f"Comment Text: {input_with_prediction[0]}")
...     print(f"Generated Output: {input_with_prediction[1][0]}")
...     print("\n\n")
... 
Comment Text: THEY DONT HUNT US!!! They could, but they don't. Reading that the freakin' idiot biologists went shooting crazy and killed 4 BEARS just because...Jesus Christ!! I am so sorry about the young man killed by a bear. He was out running in a wilderness area where, honestly, no one should be running without weapons and a couple of bear dogs. Ive been up here a long time. My daughter was going to participate in the mountain all night relay up at Hilltop that horrifying night when that lovely young teenage girl was attacked while riding her mountain bike through known bear hang outs in the middle of the NIGHT! I honestly consider it stupid for this young man to have been alone in bear country!!!!! I know his parents are feeling guilty for allowing him to do this race. I know they are. For the biologists, who always seem to over react up here in Alaska, for them to go out and kill every damn bear up there, is lunacy, but typical. The whole thing is sickening and so sad.
Generated Output: nobodyzЉ nobody nobody nobody nobody nobody nobody nobody nobody yesWЉ yes 
 yes ЉЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yesЉ yes yes yes
```

Note that the output may be non-sensical because the model couldn't complete fine-tuning to completion, but at least the existing checkpoint is usable and it early stops training to prevent wastage of time. From here, corrective action can be taken, such as lowering the learning rate, changing learning rate decay, increasing/decreasing dropout, modifying the size of the dataset, etc.

### CODE REVIEW

**Comment by arnavgarg1** [comment]
@justinxzhao Any suggestions on a good way to write an integration test for this? I spent quite a while tinkering around and trying different things but it's pretty hard to write a test that works as intended (but individually, I can see that the code does the right thing)

**Comment by justinxzhao** [comment]
> @justinxzhao Any suggestions on a good way to write an integration test for this? I spent quite a while tinkering around and trying different things but it's pretty hard to write a test that works as intended (but individually, I can see that the code does the right thing)

@arnavgarg1 If setting an obscenely large learning rate doesn't work, I see the potential conundrum here since the same `.train()` call with a `model_resume_path` loads the model weights in-method.

Here's a different idea -- use `.load()` to reload a previously trained model, make the surgical `NaN` weight change manually, and then call `.train()` to resume training, but without the `model_resume_path`. WDYT?

```python
# usual training for N steps.
output_dir = model.train()

# reload model.
model2 = LudwigModel.load(output_dir)

# surgical change.
model2.model.model…weight = float("NaN")

# resume training with more steps (change the internal config object?).
model2.train()

# Inspect output_dir artifacts, check training steps is still N 
# because the presence of the NaN weight causes training to quit.
# Another option would be to check callbacks
```

These tests may be helpful to reference:
- https://github.com/ludwig-ai/ludwig/blob/master/tests/integration_tests/test_model_training_options.py#L143
- https://github.com/ludwig-ai/ludwig/blob/master/tests/integration_tests/test_experiment.py#L810
- https://github.com/ludwig-ai/ludwig/blob/master/tests/integration_tests/test_api.py#L436

As an aside, I don't think a rigorous integration test beyond the basic unit tests you've already written is totally necessary, but I will leave that up to you!

File: tests/integration_tests/test_trainer.py [hunk 1]
```diff
@@ -417,3 +417,53 @@ def test_enable_gradient_checkpointing(tmpdir, caplog):
     # Check that the warning is emitted when the model does not support gradient checkpointing
     # but does not prevent training from starting.
     assert "Gradient checkpointing is currently only supported for model_type: llm. Skipping..." in caplog.text
+
+
+# def test_early_stopping_nan_weights(tmpdir, caplog):
```
Comment (geoffreyangus) [inline tests/integration_tests/test_trainer.py:422]: Does this need to be added back in? If not, might be good to delete it.

**Review by geoffreyangus — APPROVED** [review]

File: ludwig/trainers/trainer.py [hunk 2]
```diff
@@ -996,6 +997,13 @@ def train(
                             f"Epoch {progress_tracker.epoch} took: "
                             f"{time_utils.strdelta((time.time() - start_time) * 1000.0)}."
                         )
+
+                    # Skip saving progress if we're not saving the model. We should do this so as to not overwrite the
+                    # best model checkpoint from the previous round of evaluation so that the previous best model
+                    # weights can be used for inference instead of the current weights which are in a bad state.
+                    if has_nan_or_inf_tensors:
```
[the remark this answers is not in the document]
Comment (arnavgarg1) [reply ludwig/trainers/trainer.py:1027]: Added a logging message here: https://github.com/ludwig-ai/ludwig/pull/3740/files/804ef3bd2c1c90cff1231ec01f11e34793db3b4d#diff-86783a17ee3a22d54149070a96dc6d5d2a2387544e99d5d3af68e7374d996b57R1272

File: ludwig/trainers/trainer.py [hunk 3]
```diff
@@ -1035,7 +1043,7 @@ def train(
             if test_summary_writer is not None:
                 test_summary_writer.close()
 
-            if not self.skip_save_model and self.skip_all_evaluation:
+            if not self.skip_save_model and self.skip_all_evaluation and not has_nan_or_inf_tensors:
```
[the remark this answers is not in the document]
Comment (arnavgarg1) [reply ludwig/trainers/trainer.py:1069]: I think yes - there's no point of continuing training for now since subsequent epochs will have NaN loss. In the future, we can think of other mechanisms, such as lowering learning rate, decreasing the batch size / gradient accumulation steps, etc from the last valid checkpoint!

**Review by Infernaught — APPROVED** [review]
Nonblocking (mostly just a question): Was there a specific reason why the name was changed from "has" to "contains"?

Otherwise, LGTM!

**Comment by arnavgarg1** [comment]
@Infernaught I just realized that my function and my variable were both called the same thing which could be confusing to read through, so I decided to rename the function and keep the variable the same.
